### PR TITLE
Improve the UX of the login fallback when using SSO

### PR DESCRIPTION
**Pull request by babolivier**
* Don't show the login forms if we're currently logging in with a
  password or a token.
* Submit directly the SSO login form, showing only a spinner to the
  user, in order to eliminate from the clunkiness of SSO through this
  fallback.

### CODE REVIEW

File: synapse/static/client/login/js/login.js [hunk 1]
```diff
@@ -46,20 +49,37 @@ var setFeedbackString = function(text) {
...
+            // because, if there is a token, and this function is run, it means an error
+            // happened, and in this case it's nicer to show the form with an error
+            // rather than redirect immediately to the SSO portal.
+            $("#sso_form").show();
+        } else {
+            // Submit the SSO form instead of displaying it. The reason behind this
+            // behaviour is that the user will likely arrive here after clicking on a
+            // button, in the client, with a label such as "Continue with SSO". And even
+            // if that's not the case, it kind of makes sense to direct the user directly
+            // to the SSO portal and skip the password login form.
```
Comment (richvdh) [inline synapse/static/client/login/js/login.js]: surely we shouldn't break password login for servers that support it? We should only skip straight to SSO if that's the only option supported by the server.

File: synapse/static/client/login/js/login.js [hunk 2]
```diff
@@ -46,20 +49,37 @@ var setFeedbackString = function(text) {
...
 
 var show_login = function() {
-    $("#loading").hide();
-
     var this_page = window.location.origin + window.location.pathname;
     $("#sso_redirect_url").val(this_page);
 
-    if (matrixLogin.serverAcceptsPassword) {
-        $("#password_flow").show();
+    if (matrixLogin.serverAcceptsCas) {
```
Comment (richvdh) [inline synapse/static/client/login/js/login.js]: we can probably drop the cas stuff btw. It's superceded by SSO.

File: synapse/static/client/login/js/login.js (hunk 1 above)
Comment (babolivier) [reply synapse/static/client/login/js/login.js]: Hmm, the goal of that was to match the UX of Riot clients, but I guess that's fair.

Comment (babolivier) [reply synapse/static/client/login/js/login.js]: Done in https://github.com/matrix-org/synapse/pull/7152/commits/45409c2ef3aa42a08ab18246038c735a40b5d394

File: synapse/static/client/login/js/login.js (hunk 2 above)
Comment (babolivier) [reply synapse/static/client/login/js/login.js]: Done in https://github.com/matrix-org/synapse/pull/7152/commits/45409c2ef3aa42a08ab18246038c735a40b5d394

File: synapse/static/client/login/js/login.js [hunk 3]
```diff
@@ -46,23 +48,36 @@ var setFeedbackString = function(text) {
...
 var show_login = function() {
-    $("#loading").hide();
-
     var this_page = window.location.origin + window.location.pathname;
     $("#sso_redirect_url").val(this_page);
 
-    if (matrixLogin.serverAcceptsPassword) {
-        $("#password_flow").show();
+    if (matrixLogin.serverAcceptsSso) {
+        if (try_token() || matrixLogin.serverAcceptsPassword) {
```
Comment (richvdh) [inline synapse/static/client/login/js/login.js]: rather than this (which will try a second attempt to submit the token), I think it would make sense to give show_login an `inhibit_redirect` flag, which is true when it is called from `errorFunc` and false when it is called from `onLoad`.

File: synapse/static/client/login/js/login.js [hunk 4]
```diff
@@ -46,23 +48,36 @@ var setFeedbackString = function(text) {
...
+            // Also show the form if the server is accepting password login as well.
+            $("#sso_form").show();
+        } else {
+            // Submit the SSO form instead of displaying it. The reason behind this
+            // behaviour is that the user will likely arrive here after clicking on a
+            // button, in the client, with a label such as "Continue with SSO", so
+            // clicking on another button with the same semantics is a bit pointless.
+            $("#sso_form").submit();
+            return;
+        }
```
Comment (richvdh) [inline synapse/static/client/login/js/login.js]: I feel like this logic would make more sense if it it were reversed:

```js
    # If inhibit_redirect is false, and SSO is the only supported login method, we can
    # redirect straight to the SSO page
    if (!inhibit_redirect && !matrixLogin.serverAcceptsPassword) {
            $("#sso_form").submit();
            return;
    }

    # otherwise, show the SSO form the form
    $("#sso_form").show();
```

File: synapse/static/client/login/js/login.js [hunk 5]
```diff
@@ -46,23 +48,36 @@ var setFeedbackString = function(text) {
...
     }
 
-    if (matrixLogin.serverAcceptsSso) {
-        $("#sso_flow").show();
-    } else if (matrixLogin.serverAcceptsCas) {
-        $("#sso_form").attr("action", "/_matrix/client/r0/login/cas/redirect");
-        $("#sso_flow").show();
+    set_title(title_pre_auth);
+
+    $("#loading").hide();
```
Comment (richvdh) [inline synapse/static/client/login/js/login.js]: stick these at the end so that the logic to handle the login types isn't interrupted by unrelated code?

File: synapse/static/client/login/js/login.js [hunk 6]
```diff
@@ -47,39 +47,33 @@ var setFeedbackString = function(text) {
...
+        if (!inhibit_redirect && !matrixLogin.serverAcceptsPassword) {
             $("#sso_form").submit();
             return;
         }
+
+        // Otherwise, show the SSO form the form
+        $("#sso_form").show();
     }
 
     set_title(title_pre_auth);
```
Comment (richvdh) [inline synapse/static/client/login/js/login.js]: I think you missed this

File: synapse/static/client/login/js/login.js [hunk 7]
```diff
@@ -33,7 +33,7 @@ var submitToken = function(loginToken) {
 };
 
 var errorFunc = function(err) {
-    show_login();
+    show_login(true);
```
Comment (richvdh) [inline synapse/static/client/login/js/login.js]: comment please: why do we inhibit redirect? (so that the user can see the error, presumably)

File: synapse/static/client/login/js/login.js [hunk 8]
```diff
@@ -45,26 +49,33 @@ var setFeedbackString = function(text) {
...
-        $("#password_flow").show();
+    // If inhibit_redirect is false, and SSO is the only supported login method, we can
+    // redirect straight to the SSO page
+    if (matrixLogin.serverAcceptsSso) {
+        if (!inhibit_redirect && !matrixLogin.serverAcceptsPassword) {
+            $("#sso_form").submit();
+            return;
+        }
+
+        // Otherwise, show the SSO form the form
```
Comment (richvdh) [inline synapse/static/client/login/js/login.js]:
```suggestion
        // Otherwise, show the SSO form
```

**Comment by richvdh** [comment]
this shouldn't have gone into master. Will revert and merge to develop instead.

**Comment by babolivier** [comment]
Ugh, crap, sorry. Can't wait for develop to be the main branch. Thanks for handling.
